src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetSourcesService.cs

```diff
@@ -21,49 +21,53 @@ public sealed class NuGetSourcesService : INuGetSourcesService
         private readonly ServiceActivationOptions _options;
         private readonly IServiceBroker _serviceBroker;
         private readonly AuthorizationServiceClient _authorizationServiceClient;
-        private readonly ISharedServiceState _sharedServiceState;
+        private readonly IPackageSourceProvider _packageSourceProvider;
+
+#pragma warning disable CS0618 // Type or member is obsolete
+        private readonly IPackageSourceProvider2? _packageSourceProvider2;
+#pragma warning restore CS0618 // Type or member is obsolete
 
         public event EventHandler<IReadOnlyList<PackageSourceContextInfo>>? PackageSourcesChanged;
 
         public NuGetSourcesService(
             ServiceActivationOptions options,
             IServiceBroker serviceBroker,
             AuthorizationServiceClient authorizationServiceClient,
-            ISharedServiceState state)
+            IPackageSourceProvider packageSourceProvider)
         {
             Assumes.NotNull(serviceBroker);
             Assumes.NotNull(authorizationServiceClient);
-            Assumes.NotNull(state);
+            Assumes.NotNull(packageSourceProvider);
 
             _options = options;
             _serviceBroker = serviceBroker;
             _authorizationServiceClient = authorizationServiceClient;
-            _sharedServiceState = state;
-            _sharedServiceState.SourceRepositoryProvider.PackageSourceProvider.PackageSourcesChanged += PackageSourceProvider_PackageSourcesChanged;
+            _packageSourceProvider = packageSourceProvider;
+#pragma warning disable CS0618 // Type or member is obsolete
+            _packageSourceProvider2 = packageSourceProvider as IPackageSourceProvider2;
```

Comment:
Can we always use `IPackageSourceProvider2`. It's always available within VS no?

Reply:
Your question movivated me to finally implement https://github.com/NuGet/Home/issues/10015 / https://github.com/NuGet/Home/issues/10098